### ISSUE TITLE
Support for Qubes OS 4 in new PVH-only Xen stack

### ISSUE THREAD

**Issue by mato** [mirage/mirage#1184]
Having got back fresh from holidays, I decided to look into how hard it would be to support the existing Qubes OS 4 release in the new PVH-only Xen stack (#1159).

It turns out that Qubes ships a custom [version](https://github.com/QubesOS/qubes-vmm-xen/tree/xen-4.8) of Xen 4.8 with, among other things, _most but not quite all_ of the PVHv2 ABIs backported to it, so it looks like the changes needed are minimal.

Proceeding from @hannesm's findings in https://github.com/mirage/mirage/issues/1159#issuecomment-656197365, I have a work-in-progress branch of the Solo5 Xen bindings at https://github.com/mato/solo5/tree/support-xen-4.8, which adds support for obtaining the physical memory map from Xen via a different route if one is not present in `struct hvm_start_info`. This _appears_ to be enough to get both the Solo5 tests and `tutorial/hello` from mirage-skeleton to run on Qubes 4.0, using the following method:

1. Copy unikernel binary from a _dev_ domain to _dom0_, and create some stub files (is this needed?):
    ```
    [dom0] $ mkdir -p /var/lib/qubes/vm-kernels/mirage-test/
    [dom0] $ cd /var/lib/qubes/vm-kernels/mirage-test/
    [dom0] $ qvm-run --pass-io dev "cat /home/user/path/to/unikernel" > vmlinuz
    [dom0] $ touch modules.img
    [dom0] $ gzip -n9 < /dev/null > initramfs
    ```
2. Create a Qubes _StandaloneVM_:
    ```
    qvm-create \
        --property kernel=mirage-test \
        --property kernelopts='' \
        --property memory=64 \
        --property maxmem=64 \
        --property vcpus=1 \
        --property virt_mode=pvh \
        --label=green --class StandaloneVM \
        mirage-test
    qvm-features mirage-test no-default-kernelopts 1
    ```
3. Start the domain with:
    ```
    qvm-start mirage-test
    ```

Console logs can be found in `/var/log/xen/console/guest-test-mirage.log`.

As I have Qubes installed only on some very old (slow) spare laptop, @hannesm / @talex5 / @linse, could you try and test the work in progress branch with something more substantial, such as qubes-mirage-firewall? All you need to do is build against the package universe in #1159, with an additional pin of `solo5-bindings-xen` to `git+https://github.com/mato/solo5.git#support-xen-4.8`.

@marmarek: I have a question -- I wanted to run the Solo5 automated tests on Qubes 4, but for some reason I can't figure out a working `xl.cfg` that will correctly launch the unikernel. Just using `type=pvh` doesn't work, so it seems that part of the PVHv2 changes has not been backported. Using the following (Xen 4.8-style):

```
name = 'test_hello'
kernel = 'test_hello.xen'
pvh=1
builder='generic'
memory = 32
```

results in:

```
[mato@dom0 ~]$ sudo xl -vvv create -c test_hello.xl
Parsing config from test_hello.xl
libxl: debug: libxl_create.c:1667:do_domain_create: ao 0x1339d80: create: how=(nil) callback=(nil) poller=0x1339f00
libxl: debug: libxl_create.c:1024:initiate_domain_create: running bootloader
libxl: debug: libxl_bootloader.c:330:libxl__bootloader_run: no bootloader configured, using user supplied kernel
libxl: debug: libxl_event.c:686:libxl__ev_xswatch_deregister: watch w=0x133b218: deregister unregistered
domainbuilder: detail: xc_dom_allocate: cmdline="(null)", features="(null)"
libxl: debug: libxl_dom.c:735:libxl__build_pv: pv kernel mapped 0 path test_hello.xen
domainbuilder: detail: xc_dom_kernel_file: filename="test_hello.xen"
domainbuilder: detail: xc_dom_malloc_filemap    : 206 kB
domainbuilder: detail: xc_dom_boot_xen_init: ver 4.8, caps xen-3.0-x86_64 xen-3.0-x86_32p hvm-3.0-x86_32 hvm-3.0-x86_32p hvm-3.0-x86_64 
domainbuilder: detail: xc_dom_parse_image: called
domainbuilder: detail: xc_dom_find_loader: trying multiboot-binary loader ... 
domainbuilder: detail: loader probe failed
domainbuilder: detail: xc_dom_find_loader: trying HVM-generic loader ... 
domainbuilder: detail: loader probe failed
domainbuilder: detail: xc_dom_find_loader: trying Linux bzImage loader ... 
domainbuilder: detail: xc_dom_probe_bzimage_kernel: kernel is not a bzImage
domainbuilder: detail: loader probe failed
domainbuilder: detail: xc_dom_find_loader: trying ELF-generic loader ... 
domainbuilder: detail: loader probe OK
xc: detail: ELF: phdr: paddr=0x100000 memsz=0x94c4
xc: detail: ELF: phdr: paddr=0x10a000 memsz=0xe240
xc: detail: ELF: memory: 0x100000 -> 0x118240
xc: detail: ELF: note: PHYS32_ENTRY = 0x101000
xc: detail: ELF: Found PVH image
xc: detail: ELF: VIRT_BASE unset, using 0
xc: detail: ELF_PADDR_OFFSET unset, using 0
xc: detail: ELF: addresses:
xc: detail:     virt_base        = 0x0
xc: detail:     elf_paddr_offset = 0x0
xc: detail:     virt_offset      = 0x0
xc: detail:     virt_kstart      = 0x100000
xc: detail:     virt_kend        = 0x118240
xc: detail:     virt_entry       = 0x1011c0
xc: detail:     p2m_base         = 0xffffffffffffffff
domainbuilder: detail: xc_dom_parse_elf_kernel: xen-3.0-x86_64: 0x100000 -> 0x118240
xc: error: panic: xc_dom_core.c:917: xc_dom_parse_image: unsupported feature requested: Invalid configuration
libxl: error: libxl_dom.c:677:libxl__build_dom: xc_dom_parse_image failed: No such file or directory
domainbuilder: detail: xc_dom_release: called
(snip)
```

As you can see above, it seems to be detecting the `PHYS32_ENTRY` from the PVHv2 ELF note correctly, but then rather unhelpfully complains about an "unsupported feature requested".

Do you know if there's a way to launch a PVHv2 kernel manually via `xl` on the Xen version shipped by Qubes?

**Comment by marmarek** [comment]
I don't really know, as we use libxl through libvirt - some `xl` patches may be missing. But I'd try `type="pvh"` in the config.

**Comment by palainp** [comment]
I can confirm the correct vm start for mirage-skeleton/hello : 
```Solo5: Xen console: port 0x2, ring @0x00000000FEFFF000
Solo5: using E820 memory map
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.6.6-2-g20196872
Solo5: Memory map: 4000 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x1eafff)
Solo5:     rodata @ (0x1eb000 - 0x222fff)
Solo5:       data @ (0x223000 - 0x2d8fff)
Solo5:       heap >= 0x2d9000 < stack < 0xfa000000
2020-08-13 19:30:58 -00:00: INF [hello] hello
2020-08-13 19:30:59 -00:00: INF [hello] hello
2020-08-13 19:31:00 -00:00: INF [hello] hello
2020-08-13 19:31:01 -00:00: INF [hello] hello
Solo5: solo5_exit(0) called
```
For qubes-mirage-fw, I got an interesting start (with the minios try, it failed at the "waiting for client..." stage) but the vm halted shortly after, I will start the investigations :
```Solo5: Xen console: port 0x2, ring @0x00000000FEFFF000
Solo5: using E820 memory map
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.6.6-2-g20196872
Solo5: Memory map: 4000 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x2ecfff)
Solo5:     rodata @ (0x2ed000 - 0x34ffff)
Solo5:       data @ (0x350000 - 0x4b6fff)
Solo5:       heap >= 0x4b7000 < stack < 0xfa000000
2020-08-13 19:36:51 -00:00: INF [qubes.rexec] waiting for client...
2020-08-13 19:36:51 -00:00: INF [qubes.gui] waiting for client...
2020-08-13 19:36:51 -00:00: INF [qubes.db] connecting to server...
2020-08-13 19:36:51 -00:00: INF [qubes.db] connected
2020-08-13 19:36:51 -00:00: INF [qubes.rexec] client connected, using protocol version 2
2020-08-13 19:36:51 -00:00: INF [unikernel] QubesDB and qrexec agents connected in 0.023 s
2020-08-13 19:36:51 -00:00: WRN [dao] QubesDB key "/qubes-ip" not (yet) present; waiting for QubesDB to change...
2020-08-13 19:36:52 -00:00: INF [qubes.db] got update: "/qubes-keyboard" = "xkb_keymap {\n\txkb_keycodes  { include \"evdev+aliases(azerty)\"\t};\n\txkb_types     { include \"complete\"\t};\n\txkb_compat    { include \"complete\"\t};\n\txkb_symbols   { include \"pc+fr(oss)+inet(evdev)\"\t};\n\txkb_geometry  { include \"pc(pc105)\"\t};\n};"
2020-08-13 19:36:52 -00:00: WRN [dao] QubesDB key "/qubes-ip" not (yet) present; waiting for QubesDB to change...
2020-08-13 19:36:52 -00:00: INF [qubes.gui] client connected (screen size: 1368x768 depth: 24 mem: 4105x)
2020-08-13 19:36:52 -00:00: INF [unikernel] GUI agent connected
2020-08-13 19:36:52 -00:00: WRN [command] << Unknown command "QUBESRPC qubes.SetMonitorLayout dom0"
Aborted
Solo5: solo5_abort() called
```

**Comment by mato** [comment]
@palainp Can you get a symbol for `0x28ce2c` in your unikernel binary? The faulting address is `0x17`, so something is definitely wrong.

**Comment by palainp** [comment]
I have an up to date qubes 4.0 and I launch the pvh vm with the @talex5 qubes-test-mirage script. The faulting instruction seem to be in the malloc function : 
```
000000000028ccd0 <malloc>:
...
  28cdf6:       89 c1                   mov    %eax,%ecx
  28cdf8:       f7 d9                   neg    %ecx
  28cdfa:       21 c8                   and    %ecx,%eax
  28cdfc:       31 c9                   xor    %ecx,%ecx
  28cdfe:       f3 0f bc c8             tzcnt  %eax,%ecx
  28ce02:       8d 44 09 09             lea    0x9(%rcx,%rcx,1),%eax
  28ce06:       4c 8d 14 c5 a0 5c 4b    lea    0x4b5ca0(,%rax,8),%r10
  28ce0d:       00 
  28ce0e:       49 8b 42 10             mov    0x10(%r10),%rax
  28ce12:       48 8b 50 10             mov    0x10(%rax),%rdx
  28ce16:       49 39 d2                cmp    %rdx,%r10
  28ce19:       0f 84 11 0c 00 00       je     28da30 <malloc+0xd60>
  28ce1f:       48 39 15 92 8e 22 00    cmp    %rdx,0x228e92(%rip)        # 4b5cb8 <_gm_+0x18>
  28ce26:       0f 87 ec 03 00 00       ja     28d218 <malloc+0x548>
  28ce2c:       48 39 42 18             cmp    %rax,0x18(%rdx)
  28ce30:       0f 85 e2 03 00 00       jne    28d218 <malloc+0x548>
...
```
0x28d218 is the callq to the abort function. If you want some more analysis I attach the unikernel.xen

**Comment by mato** [comment]
@palainp Okay, so it would seem that the `malloc` heap is getting corrupted. Hard to tell why, but given that network packet flow involves grant tables and the gnttab code has all been rewritten, that's where I'd start looking. Having said that, if the problem were there I'm surprised it's not triggered sooner since there is a bunch of other inter-domain communication happening during setup, which involves the same grant table code.

How reproducible is it? (always, sometimes, how quickly after start)

> A call to OS.MM.Heap_pages.used () ...

Both `OS.MM.Heap_pages.used` and `OS.MM.Heap_pages.total` are unimplemented in the new Xen stack and will always return `0`.

Meta: If you have new information to add, please post another comment rather than editing the previous one -- that way I get a proper notification about it.

**Comment by mato** [comment]
@marmarek 

> I don't really know, as we use libxl through libvirt - some xl patches may be missing. But I'd try type="pvh" in the config.

`type=pvh` does not work in an `xl.cfg` on Qubes, so that part of PVHv2 is probably missing. As I wrote in my previous comment, using the Xen 4.8 syntax (`builder='generic', `pvh=1`) gets further, but doesn't work either.

It'd be helpful if we could make this work, since it'd make it easier to track down problems like the one @palainp is seeing by first testing with smaller "plain Xen" unikernels, to help narrow down the cause.

**Comment by palainp** [comment]
@mato the bug is fully reproducible (it occurs each times and immediatly : the vm run for no longer than 3"). To have the same unikernel one may use the following commands in a fresh vm using the "qube stock" fedora32 template :
```
sudo dnf install opam libseccomp-devel && \
opam init && opam repo add mirage-dev-3.x+xen-pvh-via-solo5 git+https://github.com/mirage/mirage-dev.git#3.x+xen-pvh-via-solo5 && opam pin solo5-bindings-xen git+https://github.com/mato/solo5.git#support-xen-4.8 && \
opam update -yu && opam install mirage && export PATH=$PATH:/home/user/.opam/default/bin/ && \
git clone https://github.com/mirage/qubes-mirage-firewall.git && cd qubes-mirage-firewall/ && \
mirage configure -t qubes && make depend && make
```

**Comment by palainp** [comment]
Some progress : The problem may not be related to the network or arp packet received because when I remove agent_listener references in the unikernel.ml The vm runs normally ```let agent_listener = RExec.listen qrexec Command.handler in```. Thus I think there may be a race condition between agent_listener and network_listener (the Uplink part, with just Client_eth there is no problem) as if I remove one or the other there is no memory access fault. Note that it is not necessary to use them in the Lwt.choose to have the problem occurs, just init them...
agent_listener use RExec.listen, and the Uplink listener use Netif.listen.

**Comment by marmarek** [comment]
> `type=pvh` does not work in an `xl.cfg` on Qubes, so that part of PVHv2 is probably missing.

Fixed it in https://github.com/QubesOS/updates-status/issues/1997

**Comment by palainp** [comment]
> > `type=pvh` does not work in an `xl.cfg` on Qubes, so that part of PVHv2 is probably missing.
> 
> Fixed it in [QubesOS/updates-status#1997](https://github.com/QubesOS/updates-status/issues/1997)

With Qubes updated with the testing repo I can start a vm using ```sudo xl -vvv create -c qubes_firewall.xl``` (based on the config file created with ```mirage configure -t xen && make depend && make```, many thanks @marmarek ! But now the new vm seem to fail to use interdomain communication (see attached logfile).
[dom0-log.txt](https://github.com/mirage/mirage/files/5080431/dom0-log.txt)

When I xl destroy the domain, I have the following : 
```
xencall:buffer: debug: total allocations:254 total releases:254
xencall:buffer: debug: current allocations:0 maximum allocations:3
xencall:buffer: debug: cache current size:3
xencall:buffer: debug: cache hits:222 misses:3 toobig:29
```

**Comment by marmarek** [comment]
`xl create` on its own does not start supporting services (`qrexec-daemon`, `qubesdb-daemon` etc). For unikernels using them, I recommend using `qvm-start` like you did before.
Alternatively, you can start those services manually (check process list for examples), but IMO that's less convenient. Is there  some debugging option missing when using qvm-start?

**Comment by mato** [comment]
@palainp Can you retry with `ocaml-freestanding` pinned to `git+https://github.com/mato/ocaml-freestanding.git#malloc-debug`? I'd like to see if we can trigger any assertion out of malloc earlier.

> Fixed it in QubesOS/updates-status#1997

@marmarek Thanks. I'll run some tests with that and report back.

**Comment by palainp** [comment]
> @palainp Can you retry with `ocaml-freestanding` pinned to `git+https://github.com/mato/ocaml-freestanding.git#malloc-debug`? I'd like to see if we can trigger any assertion out of malloc earlier.
> 
It says ```dlmalloc.i:3228: Assertion `sz == m->topsize' failed```

**Comment by mato** [comment]
> It says dlmalloc.i:3228: Assertion `sz == m->topsize' failed

Oh, good. That confirms the hypothesis that the C malloc heap is getting corrupted. Now we "just" have to figure out what's causing it. My bet is on the gnttab / ocaml-vchan code, since netif I have tested fairly extensively with the new Xen stack.

**Comment by mato** [comment]
@marmarek I can confirm that with `vmm-xen v4.8.5-22` installed in dom0, I can run the low-level Solo5 Xen tests fine using `xl`.

Will try and run some Mirage and `ocaml-vchan` test unikernels to see if anything shows up.

**Comment by mato** [comment]
@marmarek Note that `xl create -c` gets upset if not run through `sudo`, complains about `unable to exec console client: Permission denied` and then dumps core on `^]`. Not an issue for our purposes, but you may want to note it somewhere.

**Comment by mato** [comment]
Tried load-testing mirage-skeleton/device-usage/network, and the mirage/ocaml-vchan test unikernels, no issues.

**Comment by mato** [comment]
@palainp Random thought - can you try and comment out all the "memory pressure" bits in qubes-mirage-firewall, and see if anything changes?

**Comment by palainp** [comment]
Removing Memory_pressure does not change the assertion failure (same line). For testing, I can observe that the crash appears when we use simultaneously the RExec and Netif modules (commenting out one or the other let the vm terminate normally). Is it possible to use network and vchan in the same small unikernel test ?

**Comment by palainp** [comment]
Just in case, I attach here my diff from the qubes-mirage-firewall github repo. If I comment any of the ```let *_listener = ``` inunikernel.ml everything is ok, and if both are here, I have the malloc assertion fail.
[qubes-mirage-firewall-diff.txt](https://github.com/mirage/mirage/files/5086168/qubes-mirage-firewall-diff.txt)

**Comment by palainp** [comment]
I have good and bad news : the bad one is that I still do not know what cause the malloc heap corruption, the good one is that we can prevent it by adding a delay in the unikernel (:roll_eyes:), this is my diff from the qubes-mirage-firewall repo (I'm not used to ocaml, is it a real classical sleep call ?) : 
```
$ git diff
diff --git a/unikernel.ml b/unikernel.ml
index 72f2c83..b6fea4e 100644
--- a/unikernel.ml
+++ b/unikernel.ml
@@ -47,6 +47,7 @@ module Main (R : Mirage_random.S)(Clock : Mirage_clock.MCLOCK) = struct
     GUI.connect ~domid:0 () |> watch_gui;
     let qubesDB = DB.connect ~domid:0 () in
 
+    OS.Time.sleep_ns (6.0 *. 1e9 |> Int64.of_float) >>= fun () ->
     (* Wait for clients to connect *)
     qrexec >>= fun qrexec ->
     let agent_listener = RExec.listen qrexec Command.handler in
```
By now I think that this may not be a race condition but a initalization not fully completed when using Netif and RExec simultaneously. With the vm up, I can correctly use it as a netvm and ping and dns outside of my laptop, but after some times the firewall halts with this message (and it cannot start correctly if the vm behind is still running...) : 
```
Fatal error: exception (Failure "Netif: ERROR response")
Raised at file "string.ml", line 115, characters 19-34
Called from file "src/sexp.ml", line 113, characters 13-47
Solo5: solo5_exit(2) called
```

**Comment by talex5** [comment]
> (I'm not used to ocaml, is it a real classical sleep call ?)

`OS.Time.sleep_ns` creates a promise that resolves after a period of time. It won't stop the unikernel handling other callbacks while it's waiting, but it will delay executing the instructions below it. So the qrexec handshake should continue during the pause, but it won't start reading commands from the channel.

> `Netif: ERROR response`

This comes from https://github.com/mirage/mirage-net-xen/blob/3c8f08fbef0d590aa34be44810314465faba3c6b/lib/frontend.ml#L302 and means that the other end reported a problem with our message. As it's from `frontend`, this was the firewall trying to send a frame to its sys-net.

**Comment by palainp** [comment]
> So the qrexec handshake should continue during the pause, but it won't start reading commands from the channel.

Is it a correct assumption that this delay is supposed to be done with ```qrexec >>= fun qrexec ->``` and its qubesdb counterpart ? Is so, how can I check if the underlying initialisations are correctly done ?
I can restart the crashed unikernel-fw-vm with a client attached actually running if I use a bigger sleep_ns delay, you can see below that client vm 10.137.0.5 (```[client_net] add client vif {domid=14;device_id=3} with IP 10.137.0.5```) is directly connected as fw-vm 10.137.0.50 starts.

> As it's from `frontend`, this was the firewall trying to send a frame to its sys-net.

I have this error message when I start using network (even small traffic), here is the log for ```wget https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.8.2.tar.sign``` : 
```
Solo5: Xen console: port 0x2, ring @0x00000000FEFFF000
Solo5: using E820 memory map
            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Bindings version v0.6.6-2-g20196872
Solo5: Memory map: 64 MB addressable:
Solo5:   reserved @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x2eefff)
Solo5:     rodata @ (0x2ef000 - 0x352fff)
Solo5:       data @ (0x353000 - 0x4b9fff)
Solo5:       heap >= 0x4ba000 < stack < 0x4000000
2020-08-20 14:02:22 -00:00: INF [qubes.rexec] waiting for client...
2020-08-20 14:02:22 -00:00: INF [qubes.gui] waiting for client...
2020-08-20 14:02:22 -00:00: INF [qubes.db] connecting to server...
2020-08-20 14:02:22 -00:00: INF [qubes.db] connected
2020-08-20 14:02:22 -00:00: INF [qubes.rexec] client connected, using protocol version 2
2020-08-20 14:02:22 -00:00: INF [qubes.db] got update: "/mapped-ip/10.137.0.5/visible-ip" = "10.137.0.5"
2020-08-20 14:02:22 -00:00: INF [qubes.db] got update: "/mapped-ip/10.137.0.5/visible-gateway" = "10.137.0.50"
2020-08-20 14:02:33 -00:00: INF [qubes.db] got update: "/qubes-keyboard" = "xkb_keymap {\n\txkb_keycodes  { include \"evdev+aliases(azerty)\"\t};\n\txkb_types     { include \"complete\"\t};\n\txkb_compat    { include \"complete\"\t};\n\txkb_symbols   { include \"pc+fr(oss)+inet(evdev)\"\t};\n\txkb_geometry  { include \"pc(pc105)\"\t};\n};"
2020-08-20 14:02:33 -00:00: INF [qubes.gui] client connected (screen size: 1368x768 depth: 24 mem: 4105x)
2020-08-20 14:02:33 -00:00: INF [unikernel] GUI agent connected
2020-08-20 14:02:38 -00:00: INF [unikernel] QubesDB and qrexec agents connected in 16.000 s
2020-08-20 14:02:38 -00:00: INF [dao] Got network configuration from QubesDB:
            NetVM IP on uplink network: 10.137.0.3
            Our IP on uplink network:   10.137.0.50
            Our IP on client networks:  10.137.0.50
2020-08-20 14:02:38 -00:00: INF [net-xen frontend] connect 0
2020-08-20 14:02:38 -00:00: INF [net-xen frontend] create: id=0 domid=1
2020-08-20 14:02:38 -00:00: INF [net-xen frontend]  sg:true gso_tcpv4:true rx_copy:true rx_flip:false smart_poll:false
2020-08-20 14:02:38 -00:00: INF [net-xen frontend] MAC: 00:16:3e:5e:6c:00
2020-08-20 14:02:38 -00:00: INF [ethernet] Connected Ethernet interface 00:16:3e:5e:6c:00
2020-08-20 14:02:38 -00:00: INF [ARP] Sending gratuitous ARP for 10.137.0.50 (00:16:3e:5e:6c:00)
2020-08-20 14:02:38 -00:00: INF [ARP] Sending gratuitous ARP for 10.137.0.50 (00:16:3e:5e:6c:00)
2020-08-20 14:02:38 -00:00: INF [udp] UDP interface connected on 10.137.0.50
2020-08-20 14:02:38 -00:00: INF [dao] Watching backend/vif
2020-08-20 14:02:38 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:02:38 -00:00: INF [client_net] add client vif {domid=14;device_id=3} with IP 10.137.0.5
2020-08-20 14:02:38 -00:00: INF [client_net] Client 14 (IP: 10.137.0.5) ready
2020-08-20 14:02:38 -00:00: INF [ethernet] Connected Ethernet interface fe:ff:ff:ff:ff:ff
2020-08-20 14:03:07 -00:00: INF [client_eth:dom14:10.137.0.5] who-has 10.137.0.50? responding with fe:ff:ff:ff:ff:ff
2020-08-20 14:03:07 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:07 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
2020-08-20 14:03:08 -00:00: INF [memory_pressure] Writing meminfo: free 0 / 0 kB (-nan %)
Fatal error: exception (Failure "Netif: ERROR response")
Raised at file "string.ml", line 115, characters 19-34
Called from file "src/sexp.ml", line 113, characters 13-47
Solo5: solo5_exit(2) called
```

**Comment by talex5** [comment]
> My bet is on the gnttab / ocaml-vchan code, since netif I have tested fairly extensively with the new Xen stack.

@mato : did you test both the frontend and the backend sides of netif? I think the frontend only exports pages of memory, but doesn't test the import logic. The firewall acts as both a frontend and a backend, which will exercise more code. Also, the qrexec protocol IIRC uses vchans in both directions, which will also exercise both types of grant.

>> So the qrexec handshake should continue during the pause, but it won't start reading commands from the channel.
>
>Is it a correct assumption that this delay is supposed to be done with `qrexec >>= fun qrexec ->` and its qubesdb counterpart ? Is so, how can I check if the underlying initialisations are correctly done ?

Yes, that waits for the `qrexec` promise to resolve before continuing. Note that the `qrexec` on the left is a promise, but the one on the right (which has the same name and therefore shadows it) is not, and has a different type, so you can't forget to wait.

The `[qubes.rexec] client connected` message is printed when the handshake is done (whether you waited for it or not), at https://github.com/mirage/mirage-qubes/blob/master/lib/rExec.ml#L377.

**Comment by palainp** [comment]
Hi, I just got a bit beyond in finding the source of the malloc corruption. With this patch on https://github.com/palainp/ocaml-freestanding/commit/749c259ee09757e5d169a0a92f3fe2686694e152 the unikernel crashes with this weird logs (solo5-bindings-xen is pinned to ```git+https://github.com/mato/solo5.git#support-xen-4.8```, ocaml-freestanding is pinng with my repo, the unikernel is qubes-mirage-firewall) : 
```
dlmalloc.i:4661 malloc r=0xc1a030 @r->head=0xc1a038 value=0x22f81
dlmalloc.i:4662 malloc r=0xc1a030 @r->head=0xc1a038 value=0x22f81
dlmalloc.i:4661 malloc r=0xc1c030 @r->head=0xc1c038 value=0x20f81
dlmalloc.i:4662 malloc r=0xc1c030 @r->head=0xc1c038 value=0xffffffffffffffff
dlmalloc.i:3228: Assertion `sz == m->topsize' failed
Aborted
Solo5: solo5_abort() called
```
As one can notice, the value of the head field changes between two printf calls. This value change is directly related to the ```dlmalloc.i:3228: Assertion `sz == m->topsize' failed``` as the values compared in the check function are printed out with the printf. As I call a function for printing values, can it be possible that the stack and the heap collides now ?

**Comment by talex5** [comment]
Interesting. But I bet the compiler optimised out the read in the first `printf`. Maybe mark `r` as `volatile` or something to test? I seem to recall seeing Xen PVH return -1 for reads to unmapped memory before.

**Comment by mato** [comment]
The low-level parts required for supporting Qubes' custom Xen 4.8 are merged in Solo5/solo5#480 (pending a release). The rest of the discussion here is about imported grant mappings being done wrong in the new Mirage/Xen PVH stack, which is not Qubes-specific, and has continued in mirage/qubes-mirage-firewall#116, so I'm closing this issue as all done here.
